std-bits/table/src/main/java/org/enso/table/operations/AddGroupNumber.java

```diff
@@ -0,0 +1,122 @@
+package org.enso.table.operations;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+import org.enso.base.text.TextFoldingStrategy;
+import org.enso.table.data.column.storage.Storage;
+import org.enso.table.data.column.storage.numeric.LongStorage;
+import org.enso.table.data.column.storage.type.IntegerType;
+import org.enso.table.data.index.OrderedMultiValueKey;
+import org.enso.table.data.index.UnorderedMultiValueKey;
+import org.enso.table.data.table.Column;
+import org.enso.table.problems.ColumnAggregatedProblemAggregator;
+import org.enso.table.problems.ProblemAggregator;
+import org.enso.table.util.ConstantList;
+
+public class AddGroupNumber {
+  public static Storage<?> numberGroupsUnique(
+      long numRows,
+      long start,
+      long step,
+      Column[] groupingColumns,
+      ProblemAggregator problemAggregator) {
+    if (groupingColumns.length == 0) {
+      throw new IllegalArgumentException("At least one grouping column is required.");
+    }
+
+    var groupNumberIterator = new StepIterator(start, step);
+
+    long[] numbers = new long[(int) numRows];
+
+    Storage<?>[] groupingStorages =
+        Arrays.stream(groupingColumns).map(Column::getStorage).toArray(Storage[]::new);
+    ColumnAggregatedProblemAggregator groupingProblemAggregator =
+        new ColumnAggregatedProblemAggregator(problemAggregator);
+    List<TextFoldingStrategy> textFoldingStrategy =
+        ConstantList.make(TextFoldingStrategy.unicodeNormalizedFold, groupingStorages.length);
+    Map<UnorderedMultiValueKey, Long> groupNumbers = new HashMap<>();
+
+    for (int i = 0; i < numRows; i++) {
+      var key = new UnorderedMultiValueKey(groupingStorages, i, textFoldingStrategy);
+      key.checkAndReportFloatingEquality(
+          groupingProblemAggregator, columnIx -> groupingColumns[columnIx].getName());
+      var groupNumber = groupNumbers.computeIfAbsent(key, k -> groupNumberIterator.next());
+      numbers[i] = groupNumber;
+    }
+
+    return new LongStorage(numbers, IntegerType.INT_64);
+  }
+
+  public static Storage<?> numberGroupsEqualCount(
+      long numRows,
+      int groupCount,
+      long start,
+      long step,
+      Column[] orderingColumns,
+      int[] directions,
+      ProblemAggregator problemAggregator) {
+    long[] numbers = new long[(int) numRows];
+
+    var equalCountGenerator = new EqualCountGenerator(start, step, numRows, groupCount);
+
+    if (orderingColumns.length == 0) {
+      for (int i = 0; i < numRows; ++i) {
+        numbers[i] = equalCountGenerator.next();
+      }
+    } else {
+      Storage<?>[] orderingStorages =
+          Arrays.stream(orderingColumns).map(Column::getStorage).toArray(Storage[]::new);
+      List<OrderedMultiValueKey> keys =
+          new ArrayList<>(
+              IntStream.range(0, (int) numRows)
```

Comment:
I think we are doing casts like this all over the place, but I always feel a bit uneasy, shall we prefer `Math.toIntExact`?

Reply:
Done